File: matlab/attenuation_youngs/RT_param_outer.m
function [damping,kice,kwtr,int_adm,NDprams,...
            alp_scat,modT,argR,argT] =...
               RT_param_outer(om,h,ice_prams,guess)

do_test  = 0;
if nargin==0
   h     = 2;
   om    = 2*pi/10;
   H_nd  = 10;%%inf depth
   %%
   ice_prams   = fn_fill_iceprams();
   if 0
      %% test R-P drag
      ice_prams.drag_rp          = 13;
      ice_prams.viscoelastic_ws  = 0;
   elseif 1
      %% test W-S viscoelastic
      ice_prams.drag_rp          = 0;
      ice_prams.viscoelastic_ws  = 1;
   elseif 1
      %% no damping
      ice_prams.drag_rp          = 0;
      ice_prams.viscoelastic_ws  = 0;
   end
   guess       = om^2/ice_prams.g;
   do_test     = 1;
end

rho      = ice_prams.rhoice/ice_prams.rhowtr;
D        = ice_prams.young*h^3/12/(1-ice_prams.poisson^2);
L        = ( D/ice_prams.rhowtr./om.^2 ).^.2;
Lc       = ( D/ice_prams.rhowtr/ice_prams.g ).^.25;
alp_nd   = om.^2/ice_prams.g.*L;
alp      = om.^2/ice_prams.g;
alp_nd   = alp.*L;
h_nd     = h./L;
zeta_nd  = rho*h_nd;
%tst_ah   = {alp_nd,h_nd}
%%
if ~exist('H_nd')
   H_nd  = 4;%%inf depth
end
NDprams  = [alp_nd,zeta_nd,L,H_nd+0*om,Lc+0*om];

if ~exist('guess')
   guess = min(alp);
end

kice        = 0*om;
kwtr        = 0*om;
int_adm     = 0*om;
coeff_del   = 0*om;
coeff_D     = 0*om;
[dum,JJ]    = sort(om);

for j_=1:length(om)
   j  = JJ(j_);

   %%get wavenumber for ice;
   varpi = 1/alp_nd(j)-zeta_nd(j);
   [ki,BG2,coeff_del(j),coeff_D(j)]   =...
      fn_root_ice(varpi,H_nd,guess*L(j));

   kice(j)  = ki/L(j);
   guess    = kice(j);%%make guess the last root found;

   %%get wavenumber for water;
   varpi    = 1/alp_nd(j);
   Hw_nd    = H_nd+zeta_nd(j);
   [kw,BG1] = fn_root_wtr(varpi,Hw_nd,alp_nd(j));
   kwtr(j)  = kw/L(j);

   %tstBG1_  = {BG1,H_nd+zeta_nd,1/alp_nd, zeta_nd,1,1,ki,L}
   %tstBG2_  = {BG2,H_nd,1/alp_nd, zeta_nd,0,0,kw,L}
   %%get intrinsic admittance;
   %%|R|^2+int_adm*|T|^2=1
   int_adm(j)  = BG1/BG2;

   if 0
      period   = 2*pi/om(j)
      h
      wn_ice   = kice(j)
      wn_wtr   = kwtr(j)
   end
end

%% get attenuation due to Robinson-Palmer drag:
%% * dimensional eqn:
%% [ (D/rhow/om^2)*k^4+ g/om^2-rhoi*h/rhow-i*drag_rp/(om*rhow) ]*k*tanh(k*H) -1 = 0
%%    >> -i*om*drag_rp/(rhow*om^2) = -i*om*drag_rp/(rhow*om)
%% * non-dimensional eqn K=k*L:
%% [ D/(rhow*om^2*L^5)*K^4+ del-i*drag_rp/(L*om*rhow) ]*K*tanh(K*H/L) -1 = 0
%% d(del)   = -i*drag_rp/(L*om*rhow) = -i*drag_rp_nd
%% d(k)     = coeff_del*d(del)=-i*drag_rp_nd*coeff_del = i*damping_nd;
drag_rp_nd  = ice_prams.drag_rp/ice_prams.rhowtr./(om.*L);

%% W&S (still non-dim):
%% d(D)  = [-i*om*h^3*rhoi*viscoelastic_ws/6/(1+nu)]/[rhow*om^2*L^5]
%%       = [-i*om*h^3*rhoi*viscoelastic_ws/6/(1+nu)]/[rhow*om^2*L^5]
%%       = -i*h^3*rho*viscoelastic_ws/(6*(1+nu)*om*L^5)
%%       = -i*vews_nd
%% d(k)  = coeff_D*d(D)=-i*coeff_D*vews_nd = i*damping_nd;
vews_nd     = h^3*rho*ice_prams.viscoelastic_ws./(6*(1+ice_prams.poisson)*om.*L.^5);

%% add both effects together
damping_nd  = -coeff_del.*drag_rp_nd - coeff_D.*vews_nd;
damping     = damping_nd./L;

%%Get other quantities via interpolation;
alp_scat = NaN*om;
modT     = NaN*om;
argR     = NaN*om;
argT     = NaN*om;

if (nargout>5) | do_test

   for j=1:length(om)
      %% faster version with no loading of files
      %% (everything is hard-coded);
      [alp_scat(j),modT(j),argR(j),argT(j)]  =...
         RT_param_fast_v2(alp_nd(j),h_nd(j),int_adm(j));
   end

end

if do_test
   disp(['damping = ',num2str(damping, '%.9e')]);
   disp(['kice    = ',num2str(kice,    '%.9e')]);
   disp(['kwtr    = ',num2str(kwtr,    '%.9e')]);
   disp(['int_adm = ',num2str(int_adm, '%.9e')]);
   disp(' ');
   disp(['ac      = ',num2str(alp_scat,'%.9e')]);
   disp(['|T|     = ',num2str(modT,    '%.9e')]);
   disp(['Arg[R]  = ',num2str(argR,    '%.9e')]);
   disp(['Arg[T]  = ',num2str(argT,    '%.9e')]);

   c5    = 1-1i*vews_nd(1);
   del   = alp_nd(1)-zeta_nd(1);
   c1    = del - 1i*drag_rp_nd;
   Q     = roots([c5,0,0,0,c1,-1])/L(1);
   Qap   = kice(1)+1i*damping;
   %%
   [dummy,JJ]  = sort(abs(Q-Qap));
   tst         = [Q(JJ(1));Qap]
   %%
   K  = kice(1)*L(1);
   [tanh(K*H_nd),(K^4+alp_nd(1))*K*tanh(K*H_nd)-1]
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%  function y = calc_res(Z2,hr,gamma)
%  %% y=calc_res(Z2,hr,gamma)=Res(1/f(K),gamma_n),
%  %% where gamma_n is a root of the dispersion relation
%  %% f=1/K/tanh(KH)-(Dr*K^4+lam-mr*mu);
%  %% Z2={lam,mu,H}.
%  alp_nd   = Z2{1};
%  zeta_nd  = Z2{2};
%  H        = Z2{3};
%  Dr       = hr^3;
%  mr       = hr;
%  %%
%  Gam   = Dr*gamma.^4+1/alp_nd-mr*zeta_nd;
%  Gampr = Gam+4*Dr*gamma.^4;
%  denom = H*(Gam.^2.*gamma.^2-1)+Gampr;
%  y     = -gamma./denom;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function [ki,BG2,coeff_del,coeff_D]=fn_root_ice(del,H,guess)
%% finds the root of the ice dispersion relation nearest to 'guess'.

fac   = 1;
tol   = 1e-12;
k0    = guess;
dk    = NR_corr_term(k0,del,H,fac);
ki     = k0-dk;

while abs(dk) > tol
   k0 = ki;
   dk = NR_corr_term(k0,del,H,fac);
   ki  = k0-dk;
end

[dk,Lam,Lampr] = NR_corr_term(ki,del,H,fac);
denom          = H*(Lam.^2.*ki.^2-1)+Lampr;
res            = -ki/denom;
BG2            = Lam^2*res;

%% perturbations to get R&P drag and W&S viscoelastic effect
%% F=Lam(k,D,del)*k*tanh(k*H)-1=0:
%%  F_k   = [ (Lam*k)_k+H*( (Lam*k)^2-1 ) ]/(Lam*k);
%%  F_del = Lam_del/Lam = 1/Lam;
%%  F_D   = Lam_D/Lam   = k^4/Lam;
%% dF/dD       = -F_D/F_k;
%% dF/d(del)   = -F_del/F_k;

coeff_del   = -ki/denom;  %% dk/d(del) - non-dimensional
coeff_D     = -ki^5/denom;%% dk/d(D)   - non-dimensional

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function [kw,BG1]=fn_root_wtr(del,H,guess)
%% finds the root of the ice dispersion relation nearest to 'guess'.

fac   = 0;
tol   = 1e-12;
k0    = guess;
dk    = NR_corr_term(k0,del,H,fac);
kw    = k0-dk;

while abs(dk) > tol
   k0       = kw;
   dk       = NR_corr_term(k0,del,H,fac);
   kw       = k0-dk;
   %cvg_tst  = {dk,kw},pause
end

[dk,Lam,Lampr] = NR_corr_term(kw,del,H,fac);
denom          = H*(Lam.^2.*kw.^2-1)+Lampr;
res            = -kw/denom;
BG1            = Lam^2*res;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function [dk,Lam,Lampr] = NR_corr_term(k,del,H,fac)
%% dk=f/f_k, where f has the same zeros as of the dispersion function, 
%% is the correction term in the Newton-Rhapson method for finding zeros in f.

Lam   = fac*k.^4+del;
Lampr = 5*fac*k.^4+del;
x     = 7.5;
if real(k*H)<=x
   f  = Lam.*k.*sinh(k*H)-cosh(k*H);
   df = Lam.*(k*H).*cosh(k*H)+(Lampr-H).*sinh(k*H);
else
   f  = Lam.*k.*tanh(k*H)-1;
   df = Lam.*k*H+(Lampr-H).*tanh(k*H);
end
dk = f./df;

File: CXX/src/RTparam_outer.c
/* Standard files; */
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

/* My header files; */
#include <RTparam_fast.h>
#include <RTparam_hardcoded.h>
#include <RTparam_outer.h>

#define PI M_PI
#define EPS 1.0e-12
#define MAXIT 100


int RTparam_outer(double outputs[],double h,double om,double visc_rp,double guess,double params[])
{

   double alp_nd,h_nd,zeta_nd;
   double Hw_nd,varpi,L;
   double damping_nd,visc_rp_nd;
   double H_nd = 4.0;/* infinite depth used in scattering calculation; */
   /* */
   double ki,kw,avc,BG1,BG2;
   double E,g,rhow,rhoi,nu,rho;
   double D;

   E     = params[0];
   g     = params[1];
   rhow  = params[2];
   rhoi  = params[3];
   nu    = params[4];
   rho   = rhoi/rhow;
   /* */
   D        = E*(h*h*h)/12/(1-nu*nu);
   L        = exp( 0.2*log(D/rhow/om/om) );
   alp_nd   = om*om/g*L;
   h_nd     = h/L;
   zeta_nd  = rho*h_nd;

   /* printf("E=          %f\n",h); */

   /* get wavenumber for ice; */
   varpi = 1/alp_nd-zeta_nd;
   /* [ki,BG2,avc]   = gen_root_ice(varpi,H_nd,guess*L); */
   gen_root_ice(&ki,&BG2,&avc,varpi,H_nd,guess*L);
   /* *kice = ki/L; */
   outputs[1] = ki/L;

   /* get wavenumber for water; */
   varpi = 1/alp_nd;
   Hw_nd = H_nd+zeta_nd;
   /* [kw,BG1] = gen_root_wtr(varpi,Hw_nd,alp_nd); */
   gen_root_wtr(&kw,&BG1,varpi,Hw_nd,alp_nd);
   /* *kwtr  = kw/L; */
   outputs[2] = kw/L;

   /* get intrinsic admittance; */
   /* |R|^2+int_adm*|T|^2=1 */
   /* printf("\noutput: BG1   = %0.9e\n", BG1); */
   /* printf("\noutput: BG2   = %0.9e\n", BG2); */
   /* int_adm  = BG1/BG2; */
   outputs[3] = BG1/BG2;

   /* get viscous attenuation; */
   visc_rp_nd  = visc_rp/rhow/om/L;
   damping_nd  = avc*visc_rp_nd;
   /* *damping    = damping_nd/L; */
   outputs[0] = damping_nd/L;

   /* printf("\L   = %0.9e\n", L); */
   /* printf("\noutput: BG1   = %0.9e\n", outputs[0]); */

   /* printf("outputs[0]  %d\n",outputs[0]); */

   /* get interpolated attenuation coefficient; */
   /* RTparam_fast(ac,modT,argR,argT, */
   /*              alp_nd,h_nd,*int_adm); */

   RTparam_fast(&outputs[4],&outputs[5],&outputs[6],&outputs[7],
                alp_nd,h_nd,outputs[3]);



   /* printf("inp[0]          %d\n",params[0]); */
   /* printf("inp[1]          %d\n",params[1]); */
   /* printf("inp[2]          %d\n",params[2]); */
   /* printf("inp[3]          %d\n",params[3]); */
   /* printf("inp[4]          %d\n",params[4]); */



   /* printf("out[0]          %d\n",outputs[0]); */
   /* printf("out[1]          %d\n",outputs[1]); */
   /* printf("out[2]          %d\n",outputs[2]); */
   /* printf("out[3]          %d\n",outputs[3]); */
   /* printf("out[4]          %d\n",outputs[4]); */
   /* printf("out[5]          %d\n",outputs[5]); */
   /* printf("out[6]          %d\n",outputs[6]); */
   /* printf("out[7]          %d\n",outputs[7]); */

}

/*******************************************************************/
/* %%function [ki,BG2,avc]=gen_root_ice(del,H,guess) */
/* finds the root of the ice dispersion relation nearest to 'guess'. */
/* Newton-Rhapson method solver */
/* of f  = Lam*k*sinh(k*H)-cosh(k*H) = 0, Lam=k^4+del */
int gen_root_ice(double *ki2, double *BG2,double *avc,
                 double del, double H,double guess) {

   double fac,k0,dk,ki;
   double res,Lam,Lampr,denom;

   fac   = 1.0;
   k0    = guess;

   /* Call dispersion relation function; */
   /* dk    = NR_corr_term(k0,del,H,fac); */
   NR_corr_term(&dk,&Lam,&Lampr,k0,del,H,fac);
   ki = k0-dk;

   while(fabs(dk) > EPS) {
     k0 = ki;

     /* Call dispersion relation function; */
     /* dk = NR_corr_term(k0,del,H,fac); */
     NR_corr_term(&dk,&Lam,&Lampr,k0,del,H,fac);
     ki = k0-dk;
   }

   /* Call dispersion relation function; */
   /* [dk,Lam,Lampr] = NR_corr_term(ki,del,H,fac); */
   NR_corr_term(&dk,&Lam,&Lampr,ki,del,H,fac);
   /* */
   denom = H*(Lam*Lam*ki*ki-1)+Lampr;
   res   = -ki/denom;

   /* Outputs; */
   *ki2  = ki;
   *BG2  = Lam*Lam*res;
   *avc  = ki/denom; /* -1i*dk/d(visc_rp) - non-dimensional */
}

/*******************************************************************/
/* %%function [kw,BG1]=gen_root_wtr(del,H,guess) */
/* finds the root of the water dispersion relation nearest to 'guess'. */
/* Newton-Rhapson method solver */
/* of f  = del*k*sinh(k*H)-cosh(k*H) = 0 */
int gen_root_wtr(double *kw2, double *BG1,
                 double del, double H,double guess) {

   double fac,k0,dk,kw;
   double res,Lam,Lampr,denom;

   fac   = 0.0;
   k0    = guess;

   /* Call dispersion relation function; */
   /* dk    = NR_corr_term(k0,del,H,fac); */
   NR_corr_term(&dk,&Lam,&Lampr,k0,del,H,fac);
   kw = k0-dk;

   while(fabs(dk) > EPS) {
      k0       = kw;

      /* Call dispersion relation function; */
      /* dk       = NR_corr_term(k0,del,H,fac); */
      NR_corr_term(&dk,&Lam,&Lampr,k0,del,H,fac);
      kw       = k0-dk;
   }

   /* Call dispersion relation function; */
   /* [dk,Lam,Lampr] = NR_corr_term(kw,del,H,fac); */
   NR_corr_term(&dk,&Lam,&Lampr,kw,del,H,fac);
   denom          = H*(Lam*Lam*kw*kw-1)+Lampr;
   res            = -kw/denom;

   /* printf("(denom,res,Bg2)=(%.9e,%.9e,%.9e)",res,denom,Lam*Lam*res); */

   /* Outputs; */
   *kw2  = kw;
   *BG1  = Lam*Lam*res;
}

/* function [dk,Lam,Lampr] = NR_corr_term(k,del,H,fac) */
/* %% dk=f/f_k, where f has the same zeros as of the dispersion function, */
/* %% is the correction term in the Newton-Rhapson method for finding zeros in f. */
int NR_corr_term(double *dk,double *Lam2,double *Lampr2,
                 double k, double del, double H,double fac) {

   double f,df;
   double Lam,Lampr,x,k4;

   k4    = k*k*k*k;
   Lam   = fac*k4+del;
   Lampr = 5*fac*k4+del;
   x     = 7.5;

   if(fabs(k*H)<=x) {
      f  = Lam*k*sinh(k*H)-cosh(k*H);
      df = Lam*(k*H)*cosh(k*H)+(Lampr-H)*sinh(k*H);
   }
   else {
      f  = Lam*k*tanh(k*H)-1;
      df = Lam*k*H+(Lampr-H)*tanh(k*H);
   }

   /* Outputs; */
   *dk      = f/df;
   *Lam2    = Lam;
   *Lampr2  = Lampr;
}
